server/rest/src/main/java/org/infinispan/rest/InfinispanRequest.java

```diff
@@ -0,0 +1,262 @@
+package org.infinispan.rest;
+
+import java.util.List;
+import java.util.Optional;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.http.FullHttpRequest;
+import io.netty.handler.codec.http.HttpHeaderNames;
+import io.netty.handler.codec.http.QueryStringDecoder;
+import io.netty.handler.codec.http2.HttpConversionUtil;
+
+/**
+ * Representation of a HTTP request tailed for Infinispan-specific requests.
+ *
+ * @author Sebastian Łaskawiec
+ */
+public class InfinispanRequest {
+
+   private final FullHttpRequest request;
+   private final Optional<String> streamId;
+   private final ChannelHandlerContext nettyChannelContext;
+   private final Optional<String> cacheName;
+   private final String context;
+   private final Optional<String> key;
+   private final QueryStringDecoder queryStringDecoder;
+
+   private InfinispanRequest(FullHttpRequest request, ChannelHandlerContext ctx) {
+      this.request = request;
+      queryStringDecoder = new QueryStringDecoder(request.uri());
+      streamId = Optional.ofNullable(request.headers().get(HttpConversionUtil.ExtensionHeaderNames.STREAM_ID.text()));
+      this.nettyChannelContext = ctx;
+
+      //we are parsing /context/cacheName/key
+      //not that the first one is empty!
+      String[] pathParams = queryStringDecoder.path().split("/");
```

Comment:
Possible optimization: do not use split, but just iterate through the string and collect parts as you go.